activerecord/lib/active_record/counter_cache.rb

```diff
@@ -180,14 +181,24 @@ def decrement_counter(counter_name, id, by: 1, touch: nil)
       def counter_cache_column?(name) # :nodoc:
         _counter_cache_columns.include?(name)
       end
+
+      def load_schema # :nodoc:
+        super
+
+        _reflections.each do |name, reflection|
+          next unless reflection.belongs_to? && reflection.counter_cache_column
+
+          self.counter_cached_association_names |= [name.to_sym]
+        end
```

Comment:
OK, so part of this is a bit nitpick territory.

First:

```suggestion
        counter_caches = _reflections.compact_map do |name, reflection|
          next unless reflection.belongs_to? && reflection.counter_cache_column

          name.to_sym
        end
        self.counter_cached_association_names |= counter_caches
```

Writing into `class_attribute` is slow, as it's basically a `define_method`. So if you can batch it, it's best.

Then, if we trigger this in all classes, it's likely the child will have exactly the same list as the parent, in such case we'd ideally not write at all so we only hold a single array inherited from the parent:

```suggestion
        counter_caches = _reflections.compact_map do |name, reflection|
          if reflection.belongs_to? && reflection.counter_cache_column
            name.to_sym
          end
        end
        
        counter_caches |= counter_cached_association_names
        
        if counter_cached_association_names != counter_caches
          self.counter_cached_association_names = counter_caches
        end
```

Again, most of that is a bit nitpicky, as you you have enough counter caches for this to make much of a difference, you have bigger problem I think.

But I think the first snippet I shared is somewhat clearer, so to improve boottime very so slithly and for readability I think it would make sense.

Reply:
Thanks, I went with the first option 👍